walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,98 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    //If there are no events or no one is attending the event, 
+    // all requests are possible if they are within the range of the whole day.
+    if(request.getDuration() <= TimeRange.WHOLE_DAY.duration()){
+      if(events.isEmpty() || request.getAttendees().isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    } else {
+      // Otherwise no available times.
+      return Arrays.asList();
+    }
+
+    // List of existing bookedEvents for attendees in meeting request.
+    List<Event> bookedEvents = new ArrayList<>();
+
+    // Find all events that attendees in the request are already booked for.
+    for(Event event : events){
+      for (String attendee : request.getAttendees()){
+        if(event.getAttendees().contains(attendee) && !bookedEvents.contains(event)){
+          bookedEvents.add(event);
+        }
+      }
+    }
+    
+    // List of potential free time ranges for the meeting request.
+    List<TimeRange> potentialTime = new ArrayList<>();
+
+    if(bookedEvents.isEmpty()){
+      // attendees are not booked for any events, are available for the whole day.
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } else {
+
+      Event prevEvent = bookedEvents.get(0); 
+      Event currEvent = bookedEvents.get(0);
+      for(int i = 0; i < bookedEvents.size(); i++){ 
+        prevEvent = currEvent; 
+        currEvent = bookedEvents.get(i);          
+        // If only one event - consider only timerange before and after the event.
+        if(bookedEvents.size() == 1) {
+          potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(), false));
+          potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+          break;
+        }
+
+        if(potentialTime.size() == 0) { 
+          // See if there is a slot before the first booked event and if this slot.
+          // is large enough for the request. 
+          if(currEvent.getWhen().start() != TimeRange.START_OF_DAY) {
+            // If request duration is within slot between start of day and first event.
+            if((TimeRange.START_OF_DAY+request.getDuration()) <= currEvent.getWhen().start()){
+              potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(),false));  
+            } 
+            potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY,true));       
```

Comment:
nit: does this handle the situation where currEvent ends less than request.getDuration() from the end of the day?

Reply:
Added in this case